llvm/lib/Target/EVM/EVMStackifyCodeEmitter.cpp

```diff
@@ -0,0 +1,574 @@
+  // Validate stack.
+  assert(Emitter.getStackHeight() == static_cast<int>(CurrentStack.size()));
+  assert(CurrentStack.size() >= Call.NumArguments + (Call.CanContinue ? 1 : 0));
+
+  // Assert that we got the correct return label on stack.
+  if (Call.CanContinue) {
+    [[maybe_unused]] const auto *returnLabelSlot =
+        std::get_if<FunctionCallReturnLabelSlot>(
+            &CurrentStack.at(CurrentStack.size() - Call.NumArguments - 1));
+    assert(returnLabelSlot && returnLabelSlot->Call == Call.Call);
+  }
+
+  // Emit code.
+  const MachineOperand *CalleeOp = Call.Call->explicit_uses().begin();
+  assert(CalleeOp->isGlobal());
+  Emitter.emitFuncCall(Call.Call, CalleeOp->getGlobal(),
+                       Call.Call->getNumExplicitDefs() - Call.NumArguments -
+                           (Call.CanContinue ? 1 : 0),
+                       Call.CanContinue);
+
+  // Update stack, remove arguments and return label from CurrentStack.
+  for (size_t I = 0; I < Call.NumArguments + (Call.CanContinue ? 1 : 0); ++I)
+    CurrentStack.pop_back();
+
+  // Push return values to CurrentStack.
+  unsigned Idx = 0;
+  for (const auto &MO : Call.Call->defs()) {
+    assert(MO.isReg());
+    CurrentStack.emplace_back(TemporarySlot{Call.Call, MO.getReg(), Idx++});
+  }
+  assert(Emitter.getStackHeight() == static_cast<int>(CurrentStack.size()));
+}
+
+void EVMStackifyCodeEmitter::visitInst(const CFG::BuiltinCall &Call) {
+  size_t NumArgs = Call.Builtin->getNumExplicitOperands() -
+                   Call.Builtin->getNumExplicitDefs();
+  // Validate stack.
+  assert(Emitter.getStackHeight() == static_cast<int>(CurrentStack.size()));
+  assert(CurrentStack.size() >= NumArgs);
+  // TODO: assert that we got a correct stack for the call.
+
+  // Emit code.
+  Emitter.emitInst(Call.Builtin);
+
+  // Update stack and remove arguments from CurrentStack.
+  for (size_t i = 0; i < NumArgs; ++i)
+    CurrentStack.pop_back();
+
+  // Push return values to CurrentStack.
+  unsigned Idx = 0;
+  for (const auto &MO : Call.Builtin->defs()) {
+    assert(MO.isReg());
+    CurrentStack.emplace_back(TemporarySlot{Call.Builtin, MO.getReg(), Idx++});
+  }
+  assert(Emitter.getStackHeight() == static_cast<int>(CurrentStack.size()));
+}
+
+void EVMStackifyCodeEmitter::visitAssign(const CFG::Assignment &Assignment) {
+  assert(Emitter.getStackHeight() == static_cast<int>(CurrentStack.size()));
+
+  // Invalidate occurrences of the assigned variables.
+  for (auto &CurrentSlot : CurrentStack)
+    if (const VariableSlot *VarSlot = std::get_if<VariableSlot>(&CurrentSlot))
+      if (is_contained(Assignment.Variables, *VarSlot))
+        CurrentSlot = JunkSlot{};
+
+  // Assign variables to current stack top.
+  assert(CurrentStack.size() >= Assignment.Variables.size());
+  auto StackRange = make_range(CurrentStack.end() - Assignment.Variables.size(),
+                               CurrentStack.end());
+  auto RangeIt = StackRange.begin(), RangeE = StackRange.end();
```

Comment:
[zip](https://llvm.org/docs/ProgrammersManual.html#the-zip-functions)?

Reply:
Managed to find a way to refactor this further by using `llvm::copy`.